lms/djangoapps/badges/backends/badgr.py

```
@@ -30,8 +32,6 @@ class BadgrBackend(BadgeBackend):
 
     def __init__(self):
         super(BadgrBackend, self).__init__()
-        if not settings.BADGR_API_TOKEN:
-            raise ImproperlyConfigured("BADGR_API_TOKEN not set.")
```

Maybe retain the `ImproperlyConfigured` exception in `__init__()`, but throw it if `settings.BADGR_USERNAME` or `settings.BADGR_PASSWORD` is unset?

---

Sure, that makes sense :)

---

Just checking: is anything keeping you from proactively calling `_get_and_cache_oauth_tokens()` from `__init__()`? That way you'd always have a pre-cached set of tokens, and also you wouldn't need to separately check for any `settings` attributes — because any invocation of `settings.BADGR_*` would automagically raise `ImproperlyConfigured` if the corresponding setting is unset. Unless I'm missing something?

---

Hmm, we could call it here, sure. But I guess we'd want to rewrite it a bit then. We should still check if we already have a valid token or a usable refresh_token cached before creating new ones.

---

Yeah it's just a thought; it may not be a good idea after all. If you decide against it, though, I'd argue that you'd want to raise `ImproperlyConfigured` in `__init__()` not only on a missing username, but also on a missing password.

---

So I really liked the idea of loading the tokens here and I gave that a shot but I still wanted to keep the same checks before requests (are tokens cashed/valid, can we refresh) and ended up with a rather messy and repetitive code so I'd go with checking if we have all the necessary settings here. Does that look ok?